### DISCUSSION TITLE
Default Language Edition

### DISCUSSION THREAD

**Discussion by pbackus** [dlang/PhobosV3-Design#8] [General]
From [the forums][1]:

> The other major topic of discussion was what I've been calling the "Crippled by Default" design of editions, where the oldest edition (technically the last pre-edition release) is the default edition if no edition is specified. This poses a few challenges from an end-user standpoint, but the argument that ended up resonating was the idea that in engineering we always want to make the "right" way the default or easiest way to do something, and then provide escape hatches where necessary.
> 
> Therefore, the compiler should default to use the latest version and then provide the ability via a switch to set the edition, or lack thereof, of the modules an import path. This solves the problem of abandon-ware packages being accessible without presenting the new user with an ever more decayed version of the compiler. We want to put out best foot forward and presenting the last pre-editions release, which is constantly getting old as time passes, does not do that. No specific agreement was reached but Walter agreed to consider it.
> 
> When then moved on to a conversation about how Walter envisions editions actually working. Since none of have seen the document that Atila is working on, Walter shared his opinions on how it should work. Essentially, Walter would like to see a "hybrid" approach having edition attributes for specific experimental features, and then having a yearly "roll-up" edition that includes all the promoted features from the prior year. So if DIP1000 gets promoted to Edition 2025, then DIP1000 would be active by default in that edition and all subsequent editions without having to specifically enable it.
> 
> I did bring up that this was likely to cause another "function attribute soup" problem, but in general I wholeheartedly agree with the idea that editions should use this model, both C# and C++ both do something similar so it would be conceptually comfortable to users coming from those languages. Atila, if you're reading this, this is what Walter was thinking/hoping would appear, but since we don't have your document we were mostly left to speculate. We await your document drop with anticipation!

[1]: https://forum.dlang.org/thread/hqliwbdrimuzzworupdt@forum.dlang.org

**Comment by atilaneves** [comment]
I think the default edition should be the oldest one, otherwise I don't see how we won't break code.

**Reply by rikkimax** [reply]
One way to solve this is to throw CI runners at it.

Find out what editions work for dub packages, and which don't.

The registry can automatically add it to the config.

**Reply by pbackus** [reply]
Worth keeping in mind that dmd and dub can have different defaults.

Having dub default to the oldest edition will ensure that old dub packages continue to work. To give new users a good out-of-the-box experience, we can have `dub init` automatically add a `"languageEdition": $CURRENT` entry to newly-generated `dub.json` files.

If we do this, then dmd can default to the current edition without breaking existing dub packages, and new users who bypass dub and use dmd directly will not have their experience compromised.

**Reply by LightBender** [reply]
This is called "Crippled by Default". It is beyond a bad idea.  Crippled by Default means that the *default* language experience is that _oldest_ version of the language. This design will result in a plethora of negative outcomes in the name of scratching one (solvable by other means) itch.

The first of which is that as we progress through time, the default experience will become progressively less useful as new language capabilities remain unavailable by default. Of course they can always put the edition tag on it, but, this leads to the second problem.

The second is that it actively encourages programmers to _not_ upgrade their code to the newer editions. "Why bother, this is the default." Why upgrade to a newer edition, when the default edition is what everybody is using, by default. From what I understand based on my discussions with @WalterBright, if a dependency has an edition then the dependent project also has to compile to that edition (Editions are not forward compatible, you could not call DIP1000 methods from a pre-DIP1000 edition). Please correct if wrong. So placing an edition tag becomes a breaking change.

Third, nobody does this. C# defaults to the latest version and you have to specify older versions in the build system. Java defaults to the latest version and you have to specify older versions in the build system. C++ is slightly different, but achieves roughly  the same thing. C++ compilers have the concept of stable(C++17)/current(C++20)/unstable(C++23) and they default to C++17 for the most part, when C++23 is fully shipped, they'll roll up to C++20. But I know of no C++ compiler that defaults to C++98.

"But old broken code" Is not a sufficient reason to Cripple ourselves by Default and incur these long-term consequences.

As I discussed with Walter, in engineering you do the "Correct\Safe\Modern\Normal" thing by default, and you provide an escape hatch for the "Incorrect/Unsafe/Old/Abnormal" thing, and attempting to use an abandonware package is the textbook definition of "Incorrect/Unsafe/Old/Abnormal". This is best handled at the build tool level, by default, the build tool will use the latest edition, and if a dependency requires an older edition, it can either specify, or we can teach the build system to override the edition and use whatever edition we specify. This is how you deal with outdated code. The compiler was already going to need a compiler switch to override editions, the build system can use that on abandonware.

I am going to collect some stats on abandonware in code.dlang.org now.

**Reply by LightBender** [reply]
Data collected: https://gist.github.com/LightBender/32f7fb6bc9a579acff3ee06df8ae901a
There are 145 packages that have not been updated in at least 1 year and have more than 25 downloads in the past month.
Furthermore, if you look at the top packages, it tends to be stuff that is very stable and well maintained, it just hasn't needed maintenance in over a year.
I'd make the argument that this data suggests that the "old broken code" narrative is significantly over-sold versus reality.

**Comment by LightBender** [comment]
I think we can close this now. Atila has agreed that we should do Latest-by-Default.
